### PR TITLE
Align service resolution with Pilot's validation

### PR DESCRIPTION
**Pull request by Nino-K**
- This change makes service resolution more aligned with Pilot validation
during the converson of an instance to serviceEntry in Galley

### CODE REVIEW

**Comment by Nino-K** [comment]
@nmittler PTAL.

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go [hunk 1]
```diff
@@ -65,14 +65,19 @@ func (i *Instance) convertService(service *resource.Entry, outMeta *mcp.Metadata
...
 
 	spec := service.Item.(*coreV1.ServiceSpec)
-
-	resolution := networking.ServiceEntry_STATIC
 	location := networking.ServiceEntry_MESH_INTERNAL
 	endpoints := convertExternalServiceEndpoints(spec, service.Metadata)
+	var resolution networking.ServiceEntry_Resolution
+	// Resolution STATIC must have endpoints
+	if len(endpoints) != 0 {
+		resolution = networking.ServiceEntry_STATIC
```
Comment (nmittler) [inline galley/pkg/runtime/projections/serviceentry/converter/instance.go]: This isn't quite right ... the endpoints created above are for external services only. 

Pilot will fail validation for STATIC with no endpoints? Maybe we have to pass the existing endpoints into this function as well?

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go [hunk 2]
```diff
@@ -65,14 +65,19 @@ func (i *Instance) convertService(service *resource.Entry, outMeta *mcp.Metadata
...
+	}
 
 	// Check for an external service
+	host := serviceHostname(service.ID.FullName, i.domain)
 	externalName := ""
-	if spec.Type == coreV1.ServiceTypeExternalName && spec.ExternalName != "" {
+	err := model.ValidateFQDN(host)
+	if err == nil && spec.Type == coreV1.ServiceTypeExternalName && spec.ExternalName != "" {
```
Comment (nmittler) [inline galley/pkg/runtime/projections/serviceentry/converter/instance.go]: why is the validation result of the hostname included in this `if` condition?

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go [hunk 3]
```diff
@@ -83,7 +88,8 @@ func (i *Instance) convertService(service *resource.Entry, outMeta *mcp.Metadata
 	if spec.ClusterIP != "" && spec.ClusterIP != coreV1.ClusterIPNone {
 		addr = spec.ClusterIP
 	}
-	if addr == model.UnspecifiedIP && externalName == "" {
+	// Resolution NONE must not have endpoints
+	if len(endpoints) == 0 && addr == model.UnspecifiedIP && externalName == "" {
```
Comment (nmittler) [inline galley/pkg/runtime/projections/serviceentry/converter/instance.go]: why this change? Isn't `NONE` the new default if there are no endpoints?

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go (hunk 1 above)
Comment (Nino-K) [reply galley/pkg/runtime/projections/serviceentry/converter/instance.go]: Yeah pilot has specific validation for STATIC with no endpoints (here)[https://github.com/istio/istio/blob/master/pilot/pkg/model/validation.go#L2377]. I will look into passing the existing endpoints.

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go (hunk 2 above)
Comment (Nino-K) [reply galley/pkg/runtime/projections/serviceentry/converter/instance.go]: This is to distinguish between the STATIC and DNS. DNS resolution requires a valid FQDN.

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go (hunk 3 above)
Comment (Nino-K) [reply galley/pkg/runtime/projections/serviceentry/converter/instance.go]: I think there is still chance of qualifying for `addr == model.UnspecifiedIP && externalName == ""` while having endpoints which makes NONE invalid.

Comment (nmittler) [reply galley/pkg/runtime/projections/serviceentry/converter/instance.go]: It's not obvious that's what you're doing here.  It might be better to have a cleanup condition toward the end that does something like:

```go
if resolution = networking.ServiceEntry_NONE && len(endpoints) > 0 {
   // NONE is disallowed when there are endpoints.
    resolution = networking.ServiceEntry_STATIC
}
```

File: galley/pkg/runtime/projections/serviceentry/converter/instance.go (hunk 2 above)
Comment (Nino-K) [reply galley/pkg/runtime/projections/serviceentry/converter/instance.go]: Sure, SGTM
